keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,682 @@
+- Protect `PersistentVolumeClaim` while being used by `Pod` (Replace the background logic for pvc-protection)
+- Protect `Secret` while being used by `Pod`, `PersistentVolume`, `VolumeSnapshotContent` (Secret protection above)
+- Protect `ConfigMap` while being used by `Pod`
+- Protect `StorageClass` while being used by `PersistentVolume`
+- Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
+- Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
+- Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
+- Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
+- Protect resources that users marked as in-use
+
+### Goals
+
+- Provide a generic mechanism that prevents a resource from being deleted when it shouldn't be deleted
+
+### Non-Goals
+
+- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
+- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
+- Implement any of the above potential use cases
+
+## Proposal
+
+A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
+Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validation in api-server.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.
+
+#### Story 2
+
+A user sets `Liens` to an object to be protected by this mechanism. For example, when a user knows that `Namespace` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a protection.
+
+<!--
+### Notes/Constraints/Caveats (Optional)
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
+Users or controllers can add a string to `Liens` to ask to protect a resource.
+A newly introduced validation in api-server will reject deletion requests for resources with non-empty `Liens`.
+
+`Liens` is defined as a slice of strings, like `Finalizers`.
+The strings need to be namespaced keys.
+Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
```

Comment:
Are liens modified using spec or status or a different subresource?  When we created finalizers, we realized that the power to change a finalizer did not necessarily imply the power to change spec or status, but we never formalized that. 

I'd like to avoid repeating the same mistake with another global metadata field.  The power to create and modify liens seems independent of the ability to modify spec or status.

Reply:
In the current KEP, advisory feature model is chosen, so Liens field remains in metadata and is not defined as a subresource. If we find use cases that require hard locking feature and choose the model, making it a subresource and moving it to outside the metadata would be needed.

Reply:
> In the current KEP, advisory feature model is chosen, so Liens field remains in metadata and is not defined as a subresource.

That may not be sufficient.  Someone may want to grant a controller sufficient power to set or clear a lien, but not the power to modify the intent of the resource or the status.

Reply:
I believe creation of a subresource to allow an actor to change liens but not other parts of the object is possible as a followup. I am not sure it's mandatory for the first version.

I believe the right to edit spec should automatically include the right to edit liens, to support creating objects with liens already in place.

Reply:
> I am not sure it's mandatory for the first version.

Having been through finalizers, if it's not in the first version, I lack confidence that it will appear later.

Reply:
xref: https://github.com/kubernetes/kubernetes/issues/72637#issuecomment-630586536

Reply:
Can we pleaaaaaaaaase add a `finalize` subresource and a protocol to add/clear those?

Reply:
> Can we pleaaaaaaaaase add a finalize subresource and a protocol to add/clear those?

+1

Reply:
@deads2k @lavalamp @thockin @wojtek-t @alvaroaleman Created a separate [KEP](https://github.com/kubernetes/enhancements/pull/3286) for subresourcing finalizers. Please review it.

(Also, I'm updating this KEP).

Reply:
> Would we be able to add the feature today to finalizers or is there a design reason that makes it impossible (backward incompatibility)?

https://github.com/kubernetes/kubernetes/issues/59591 suggests we wouldn't be able to, at least not by default (which is what matters).